persist/fs/retriever.go

```diff
@@ -285,18 +321,45 @@ func (r *blockRetriever) Stream(
 	startTime time.Time,
 	onRetrieve block.OnRetrieveBlock,
 ) (xio.SegmentReader, error) {
-	reqs, err := r.shardRequests(shard)
-	if err != nil {
-		return nil, err
-	}
-
 	req := r.reqPool.Get()
 	req.shard = shard
 	req.id = id
 	req.start = startTime
 	req.onRetrieve = onRetrieve
 	req.resultWg.Add(1)
 
+	// Capture variable and RLock() because this slice can be modified in the
+	// Open() method
+	var seekerMgr FileSetSeekerManager
+	r.RLock()
+	// This should never happen
+	if len(r.seekerMgrs) < 1 {
+		r.RUnlock()
+		return nil, errNoSeekerMgrs
+	}
+	seekerMgr = r.seekerMgrs[0]
+	r.RUnlock()
+
+	// It doesn't matter which seekerManager we use (they're all identical, we
+	// just have multiple for concurrency reasons), so we use the first one
+	// because it's guaranteed to be there (concurrency cannot be < 1)
+	seeker, err := seekerMgr.Seeker(shard, startTime)
+	if err != nil {
+		return nil, err
+	}
+	// If the ID is not in the seeker's bloom filter, then it's definitely not on
+	// disk and we can return immediately
+	if !seeker.ConcurrentIDBloomFilter().Test(id.Data().Get()) {
+		// No need to call req.onRetrieve.OnRetrieveBlock if there is no data
+		req.resultWg.Done()
```

Comment:
Instead of calling `req.resultWg.Done()` directly, for better encapsulation perhaps is better instead to `req.onRetrieved(ts.Segment{})`

Reply:
yeah fair